botocore/operation.py

```diff
@@ -77,6 +80,30 @@ def call(self, endpoint, **kwargs):
                           model=self.model,
                           params=kwargs)
         request_dict = self.build_parameters(**kwargs)
+
+        service_name = self.service.service_name
```

Comment:
This is really confusing and hard to follow.  Can you explain why this code is necessary?  Also, given you're already passing in an endpoint, it has potential to be wrong because this is relying on the fact that it can construct the internals of an Endpoint the exact same as the passed in `endpoint` object, which may not be the case.